text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,392 @@
+match y {
+    destructure!(Bar { a, b: 1 }) => /* ... */,
+    _ => panic!(".b != 1")
+}
+```
+
+Here, if we were to move fields out of `Bar` from left to right, we might move `a`, then check whether `b` matches `1`, and if it doesn't we already moved `a` out. 
+
+In theory this is not impossible to implement. By making this expand to code that uses `ManuallyDrop`, and when a pattern is rejected semantically move values back. However, one can construct even more complicated scenarios than just this one, and actually defining the semantics of this operations quickly becomes incredibly complicated.
+
+Just as a small example of that, consider this:
+
+```rust
+match z {
+    destructure!(Baz { a, b: destructure!(Bar { a, b: 1 }) }) => /* ... */,
+    _ => panic!(".b.b != 1")
+}
+```
+
+Where both `Baz` and `Bar` implement `Drop`.
+
+As a final note on this, making `destructure!()` expand to a pattern on its own is possible, and even refutability is possible if we only allow "one level" of refutability (`Some(_)`, but not `Some(1)`). However, we simply do not think there are many good uses for this. 
+The one scenario we could imagine is that might be useful for `destructure!()` on `enum`s, but we also think that even using `destructure!()` on `enum`s will be very uncommon (as types implementing `Drop` are almost always `struct`s) and might not outweigh the increased implementation complexity and the fact that we need to communicate this one level matching rule to users.
+
+## Macro expression and Magic type
+
+Instead of having a macro be a statement or pattern, we could make it an expression returning a type (say `Destructure<T>`) with magic similar to that of a `Box` and `ManuallyDrop`. Specifically, it would:
+1. Allow moving `T`'s fields out of `Destructure<T>` even if `T: Drop`
+2. Not drop `T` when `Destructure<T>` is dropped
+3. But drop all fields of `T` when `Destructure<T>` is dropped (except the moved-out ones)
+4. Be able to be matched directly with a pattern that would match `T`
+
+Note that having a macro produce `Destructure<T>`, as opposed to a function is still required, since we still want to check that `T` is constructable in the current context.
+
+This would allow writing the `BufWriter::into_parts` example from [Motivation](#motivation) like this:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    let Self { buf, inner, panicked } = destructure!(self);
+    let buf = if !panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    (inner, buf)
+}
+```
+
+Or closer to the original code:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    let this = destructure!(self);
+    let buf = this.buf;
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    (this.inner, buf)
+}
+```
+
+The upside of this is that `Destructure` type can be used as input into functions which are expected to not drop `T`, such as potential `DropOwned`:
```

Comment:
A future version of Rust that had `&move` references (a kind of pointer which transfers ownership of the value, but doesn't own the memory it points to) could simply use the statement or pattern `destructure!`: 

```rust
impl DropOwned for X {
    fn drop(self: &move Self) {
        destructure!(let Self(inner) = *self);
        drop(inner);
    }
}
```

This would have the advantage of not forcing a move of the `X` value itself, and thus doesn’t change or magically fudge the ABI — in particular, the above `DropOwned::drop()` does exactly what `drop_in_place()` does (drop the entire value), whether or not the author of the code does so explicitly.

This would also align neatly with “pinned drop” (executing drop glue on a value that is statically known to be pinned), by offering a `Pin<&move Self>` instead. At that point, we have a `drop` with all the properties missing from today's `Drop::drop()`: it can move fields out, *and* it is unable to accidentally break pinning.

Reply:
..."a kind of pointer which transfers ownership of the value, but doesn't own the memory it points to"? que?

Reply:
I'm not sure what error you intend to pointing to. My understanding of the `&move` idea (which I could not find an existing writeup of to link to, or I would have) is that:

* like `Box<T>`, a `T` can be moved out of `&move T`,
* like `Box<T>`, if the `T` is not moved out, then it is dropped when `&move T` is dropped,
* unlike `Box<T>`, no heap allocation is freed (because the allocation might belong to e.g. a parent stack frame, or some other data structure).

Thus, `&move T` owns the value but not the allocation. Does that clarify?

…that said, I just realized that this is incompatible with `Pin` without also having unleakable/linear types (you mustn't deallocate a pinned value without moving it). So, my last paragraph was a bit incomplete.

Reply:
[Here](https://internals.rust-lang.org/t/a-sketch-for-move-semantics/18632/19?u=illicitonion)'s my go-to ~writeup on the topic.